tools/trace2syz/trace2syz/parseInnerCalls.go

```diff
@@ -0,0 +1,118 @@
+package trace2syz
+
+import (
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+)
+
+func parseInnerCall(syzType prog.Type, traceType *call, ctx *Context) prog.Arg {
+	switch traceType.CallName {
+	case "htons":
+		return htonsHtonl(syzType, traceType, ctx)
+	case "htonl":
+		return htonsHtonl(syzType, traceType, ctx)
+	case "inet_addr":
+		return inetAddr(syzType, traceType, ctx)
+	case "inet_pton":
+		return inetPton(syzType, traceType, ctx)
+	case "makedev":
+		return makedev(syzType, traceType, ctx)
+	default:
+		log.Fatalf("Inner Call: %s Unsupported", traceType.CallName)
+	}
+	return nil
+}
+
+func makedev(syzType prog.Type, traceType *call, ctx *Context) prog.Arg {
+	var major, minor, id int64
+
+	arg1 := traceType.Args[0].(*expression)
+	arg2 := traceType.Args[1].(*expression)
+	major = int64(arg1.Eval(ctx.Target))
+	minor = int64(arg2.Eval(ctx.Target))
+
+	id = ((minor & 0xff) | ((major & 0xfff) << 8) | ((minor & ^0xff) << 12) | ((major & ^0xfff) << 32))
+
+	return prog.MakeConstArg(syzType, uint64(id))
+
+}
+
+func htonsHtonl(syzType prog.Type, traceType *call, ctx *Context) prog.Arg {
+	if len(traceType.Args) > 1 {
+		panic("Parsing Htons/Htonl...it has more than one arg.")
+	}
+	switch typ := syzType.(type) {
+	case *prog.ProcType:
+		switch a := traceType.Args[0].(type) {
+		case *expression:
+			val := a.Eval(ctx.Target)
+			if val >= typ.ValuesPerProc {
+				return prog.MakeConstArg(syzType, typ.ValuesPerProc-1)
+			}
+			return prog.MakeConstArg(syzType, val)
+		default:
+			panic("First arg of Htons/Htonl is not expression")
+		}
+	case *prog.ConstType, *prog.IntType, *prog.FlagsType:
+		switch a := traceType.Args[0].(type) {
+		case *expression:
+			val := a.Eval(ctx.Target)
+			return prog.MakeConstArg(syzType, val)
+		default:
+			panic("First arg of Htons/Htonl is not expression")
+		}
+	default:
+		log.Fatalf("First arg of Htons/Htonl is not const Type: %s\n", syzType.Name())
+	}
+	return nil
+}
+
+func inetAddr(syzType prog.Type, traceType *call, ctx *Context) prog.Arg {
+	unionType := syzType.(*prog.UnionType)
+	var optType prog.Type
+	var innerArg prog.Arg
+	if len(traceType.Args) > 1 {
+		panic("Parsing InetAddr...it has more than one arg.")
+	}
+	switch a := traceType.Args[0].(type) {
+	case *ipType:
+		switch a.Str {
+		case "0.0.0.0":
+			optType = unionType.Fields[0]
```

Comment:
So I have two high level questions. 

1. How about I handle special unions like ```ipv4_addr``` or ```ipv6_addr``` seaparately? In this case i would check to see if the corresponding IR type is a ```Call``` like ```inet_addr``` and I select the option based on the value. Essentially, for each field in the union, if it is a const, I check to see if the result from ```ipv4_addr``` matches the const. I think this would eliminate a lot of the hidden assumptions which I am currently making. 

2. Currently parseInnerCalls.go is meant to handle cases where strace reverses an argument into the call  that generated the argument. I.e. ```inet_addr("0.0.0.0")```.  It doesn't seem like there is a good command line option to just show the result of this call which means I currently have to evaluate the call in some fashion. There are two options: 1) modify strace to just write the result of the call or 2) handle the calls in this tool. Do you have a preference or is there a better 3rd option?

Reply:
> There are two options: 1) modify strace to just write the result of the call or 2) handle the calls in this tool.

Let's start with option 2. This strace format seems to be reasonably stable and is not changing over time.

Reply:
To prove the point, these union indexes are already seems to be out of sync :)
"127.0.0.1" should be 4, and then all subsequent are off-by-one. The default value for 3 will generate something completely different.

Reply:
Re question 1
If we need a union option that contains 127.0.0.1, then I would scan this union in some target analysis function that runs before actual transformation and find it in the union (or return an error if we can't find it). Then we could test this analysis in unit-tests and ensure that whenever system call descriptions change we still can find the necessary union option, or loudly fail the test.
However, in this particular case, I think we can always just the last random option that just contains full IP value.
We could make that random int32be option the first option of the union (for both ipv4 and ipv6) and always take the first union option here (and somewhere before test that the first option actually have the type we expect).

Reply:
Oh I didn't notice the ```random``` option and haven't tried that before. What if I identify the field that gives me tightest match i.e. evaluate ```inet_addr``` to give me an int like 0x0, 0x7f000001. I then scan the fields of the union to find the tightest match, ideally a const that has the same value I evaluated. If there isn't a match, I should fallback to the next tightest field which in this case should will just be the ```random``` option since it is an int type. I think this approach should be fairly robust to changes. Agree?

Reply:
I would go with just using the random option all the time. Because why make it more complex than necessary?

Reply:
Does the random option generate a completely random ip?

Reply:
But we will fill in the right value.
That option just says "there can be any 4-byte value" and we fill in the right one.
After mutation it can be changed to a random one, but it's true for all other options too.

Reply:
That's perfect! Thanks.